File: lib/src/domain/consensus_engine.dart
/// Consensus Engine — Pure domain logic.
///
/// Combines the Micho Method (primary) with other trading method signals to
/// produce composite BUY / SELL recommendations.
///
/// **Consensus BUY (GREEN)**: Micho BUY is triggered AND at least one other
/// method also triggered a BUY signal on the same evaluation cycle.
///
/// **Consensus SELL (RED)**: Micho SELL is triggered AND at least one other
/// method also triggered a SELL signal on the same evaluation cycle.
///
/// The engine is method-agnostic. Callers provide a list of [MethodSignal]
/// results from whatever methods are active; the engine groups them by
/// direction and checks consensus rules.
library;

import 'entities.dart';
import 'micho_method_detector.dart';

/// Outcome of a consensus evaluation for one ticker.
class ConsensusResult {
  const ConsensusResult({
    required this.ticker,
    required this.buySignal,
    required this.sellSignal,
    required this.buyMethods,
    required this.sellMethods,
  });

  /// The ticker symbol.
  final String ticker;

  /// Non-null when consensus BUY is triggered.
  final MethodSignal? buySignal;

  /// Non-null when consensus SELL is triggered.
  final MethodSignal? sellSignal;

  /// All methods that triggered a BUY (including Micho if it triggered).
  final List<MethodSignal> buyMethods;

  /// All methods that triggered a SELL (including Micho if it triggered).
  final List<MethodSignal> sellMethods;

  /// `true` if at least one consensus signal (BUY or SELL) is triggered.
  bool get hasConsensus => buySignal != null || sellSignal != null;
}

class ConsensusEngine {
  const ConsensusEngine();

  /// Evaluate consensus from a flat list of method signals for a single ticker.
  ///
  /// [signals] should include Micho signals AND all other method signals.
  /// The engine does not invoke detectors — callers are responsible for
  /// running each detector first and gathering the results.
  ConsensusResult evaluate({
    required String ticker,
    required List<MethodSignal> signals,
  }) {
    final List<MethodSignal> buySignals = [];
    final List<MethodSignal> sellSignals = [];

    for (final MethodSignal signal in signals) {
      if (!signal.isTriggered) continue;
      if (_isBuyType(signal.alertType)) {
        buySignals.add(signal);
      } else if (_isSellType(signal.alertType)) {
        sellSignals.add(signal);
      }
    }

    final bool michoBuy = buySignals.any(
      (MethodSignal s) => s.alertType == AlertType.michoMethodBuy,
    );
    final bool michoSell = sellSignals.any(
      (MethodSignal s) => s.alertType == AlertType.michoMethodSell,
    );

    final int otherBuyCount = buySignals
        .where((MethodSignal s) => s.alertType != AlertType.michoMethodBuy)
        .length;
    final int otherSellCount = sellSignals
        .where((MethodSignal s) => s.alertType != AlertType.michoMethodSell)
        .length;

    final MethodSignal? consensusBuy = (michoBuy && otherBuyCount >= 1)
        ? MethodSignal(
            ticker: ticker,
            methodName: 'Consensus',
            alertType: AlertType.consensusBuy,
            isTriggered: true,
            evaluatedAt: DateTime.now(),
            currentClose: buySignals.first.currentClose,
            description:
                'Consensus BUY: Micho + '
                '$otherBuyCount other method(s) agree',
          )
        : null;

    final MethodSignal? consensusSell = (michoSell && otherSellCount >= 1)
        ? MethodSignal(
            ticker: ticker,
            methodName: 'Consensus',
            alertType: AlertType.consensusSell,
            isTriggered: true,
            evaluatedAt: DateTime.now(),
            currentClose: sellSignals.first.currentClose,
            description:
                'Consensus SELL: Micho + '
                '$otherSellCount other method(s) agree',
          )
        : null;

    return ConsensusResult(
      ticker: ticker,
      buySignal: consensusBuy,
      sellSignal: consensusSell,
      buyMethods: buySignals,
      sellMethods: sellSignals,
    );
  }

  static bool _isBuyType(AlertType type) {
    return type == AlertType.michoMethodBuy ||
        type == AlertType.rsiMethodBuy ||
        type == AlertType.macdMethodBuy ||
        type == AlertType.bollingerMethodBuy;
  }

  static bool _isSellType(AlertType type) {
    return type == AlertType.michoMethodSell ||
        type == AlertType.rsiMethodSell ||
        type == AlertType.macdMethodSell ||
        type == AlertType.bollingerMethodSell;
  }
}

File: src/domain/consensus-engine.ts
/**
 * Consensus Engine — Pure domain logic.
 * Ported from Dart: lib/src/domain/consensus_engine.dart
 *
 * Consensus BUY: Micho BUY + at least one other method BUY.
 * Consensus SELL: Micho SELL + at least one other method SELL.
 *
 * Micho method is the "approved" primary method and carries 3× weight in
 * the strength score by default. Per-method weights (G20) allow users to
 * override this — a weight of 0 disables a method from the tally.
 */
import type {
  ConsensusResult,
  ConsensusExplanation,
  MethodSignal,
  MethodName,
  MethodWeights,
  SignalDirection,
} from "../types/domain";
import { DEFAULT_METHOD_WEIGHTS } from "../types/domain";

const BUY_METHODS = new Set([
  "Micho",
  "RSI",
  "MACD",
  "Bollinger",
  "Stochastic",
  "OBV",
  "ADX",
  "CCI",
  "SAR",
  "WilliamsR",
  "MFI",
  "SuperTrend",
]);

/**
 * Resolve the effective weight for a single method.
 * Uses user-provided weights, falling back to DEFAULT_METHOD_WEIGHTS, then 1.
 */
function resolveWeight(method: string, weights?: MethodWeights): number {
  if (weights && method in weights) {
    const w = weights[method as keyof MethodWeights];
    if (w !== undefined) return w;
  }
  const d = DEFAULT_METHOD_WEIGHTS[method as keyof MethodWeights];
  return d !== undefined ? d : 1;
}

/**
 * Evaluate consensus from a list of method signals for a single ticker.
 * Accepts optional per-method weights (G20); falls back to built-in defaults
 * (Micho = 3×, all others = 1×) when omitted.
 */
export function evaluateConsensus(
  ticker: string,
  signals: readonly MethodSignal[],
  weights?: MethodWeights,
): ConsensusResult {
  const buySignals = signals.filter((s) => s.direction === "BUY" && BUY_METHODS.has(s.method));
  const sellSignals = signals.filter((s) => s.direction === "SELL" && BUY_METHODS.has(s.method));

  const michoBuy = buySignals.some((s) => s.method === "Micho");
  const michoSell = sellSignals.some((s) => s.method === "Micho");

  // Weighted vote tallies
  const michoWeight = resolveWeight("Micho", weights);
  const otherBuyWeight = buySignals
    .filter((s) => s.method !== "Micho")
    .reduce((sum, s) => sum + resolveWeight(s.method, weights), 0);
  const otherSellWeight = sellSignals
    .filter((s) => s.method !== "Micho")
    .reduce((sum, s) => sum + resolveWeight(s.method, weights), 0);

  // Total theoretical denominator = sum of all active method weights
  const totalWeighted = [...BUY_METHODS].reduce((sum, m) => sum + resolveWeight(m, weights), 0);

  let direction: SignalDirection;
  let strength: number;

  const michoEnabled = michoWeight > 0;
  const otherBuyCount = buySignals.filter((s) => s.method !== "Micho").length;
  const otherSellCount = sellSignals.filter((s) => s.method !== "Micho").length;

  if (michoEnabled && michoBuy && otherBuyCount >= 1 && otherBuyWeight > 0) {
    direction = "BUY";
    strength = totalWeighted > 0 ? (michoWeight + otherBuyWeight) / totalWeighted : 0;
  } else if (michoEnabled && michoSell && otherSellCount >= 1 && otherSellWeight > 0) {
    direction = "SELL";
    strength = totalWeighted > 0 ? (michoWeight + otherSellWeight) / totalWeighted : 0;
  } else {
    direction = "NEUTRAL";
    strength = 0;
  }

  const inputMethods = [...new Set(signals.map((signal) => signal.method))];
  const methodWeights: Partial<Record<MethodName, number>> = {};
  for (const method of BUY_METHODS) {
    methodWeights[method as keyof typeof methodWeights] = resolveWeight(method, weights);
  }
  const evaluatedAt = signals.reduce(
    (latest, signal) => (signal.evaluatedAt > latest ? signal.evaluatedAt : latest),
    "",
  );
  const limitations: string[] = ["Consensus uses technical signals and is not financial advice."];
  if (signals.length === 0) limitations.push("No method signals were available.");
  if (!michoEnabled)
    limitations.push("Micho is disabled, so BUY and SELL consensus cannot trigger.");
  if (michoEnabled && direction === "NEUTRAL") {
    limitations.push("A Micho signal and at least one active supporting method are required.");
  }

  const explanation: ConsensusExplanation = {
    evaluatedAt,
    inputMethods,
    methodWeights,
    limitations,
  };

  return {
    ticker,
    direction,
    buyMethods: buySignals,
    sellMethods: sellSignals,
    strength,
    explanation,
  };
}
